aws-lc-rs/src/cipher.rs

```diff
@@ -3,158 +3,1270 @@
+/// ];
+///
+/// let mut in_out_buffer = Vec::from(ciphertext);
+///
+/// let key = UnboundCipherKey::new(
+///     &AES_128,
+///     &[
+///         0x5b, 0xfc, 0xe7, 0x5e, 0x57, 0xc5, 0x4d, 0xda, 0x2d, 0xd4, 0x7e, 0x07, 0x0a, 0xef,
+///         0x43, 0x29,
+///     ],
+/// )?;
+/// let decrypting_key = DecryptingKey::ctr(key, context)?;
+/// let plaintext = decrypting_key.decrypt(&mut in_out_buffer)?;
+/// assert_eq!("Hello World!".as_bytes(), plaintext);
+///
+///     # Ok(())
+/// # }
+/// ```
+///
+/// ## Getting an immutable reference to an IV slice.
+///
+/// `CipherContext` implements `TryFrom<&CipherContext>` for `&[u8]` allowing immutable references
+/// to IV bytes returned from cipher encryption operations.
+///
+/// ```rust
+/// # use std::error::Error;
+/// # fn main() -> Result<(), Box<dyn Error>> {
+/// use aws_lc_rs::cipher::CipherContext;
+/// # use aws_lc_rs::cipher::{EncryptingKey, UnboundCipherKey, AES_128};
+/// # let original_message = "Hello World!".as_bytes();
+/// # let mut in_out_buffer = Vec::from(original_message);
+/// # let key_bytes: &[u8] = &[
+/// #    0x68, 0xf9, 0x46, 0x1a, 0xde, 0x8d, 0x35, 0x38, 0x7b, 0x50, 0xcc, 0x9a, 0x36, 0x64, 0xf8,
+/// #    0x9d,
+/// # ];
+/// # let key = UnboundCipherKey::new(&AES_128, key_bytes)?;
+/// # let encrypting_key = EncryptingKey::ctr(key)?;
+/// #
+/// let context: CipherContext = encrypting_key.encrypt(&mut in_out_buffer)?;
+/// let iv_bytes: &[u8] = (&context).try_into()?;
+/// assert_eq!(16, iv_bytes.len());
+/// #
+/// #    Ok(())
+/// # }
+/// ```
+///
+///
+#[non_exhaustive]
+pub enum CipherContext {
+    /// A 128-bit Initalization Vector.
+    Iv128(FixedLength<IV_LEN_128_BIT>),
+
+    /// No input to the cipher mode.
+    None,
 }
 
-unsafe impl Send for SymmetricCipherKey {}
-// The AES_KEY value is only used as a `*const AES_KEY` in calls to `AES_encrypt`.
-unsafe impl Sync for SymmetricCipherKey {}
+impl<'a> TryFrom<&'a CipherContext> for &'a [u8] {
+    type Error = Unspecified;
+
+    fn try_from(value: &'a CipherContext) -> Result<Self, Unspecified> {
+        match value {
+            CipherContext::Iv128(iv) => Ok(iv.as_ref()),
+            CipherContext::None => Err(Unspecified),
+        }
+    }
+}
 
-impl Drop for SymmetricCipherKey {
-    fn drop(&mut self) {
-        // Aes128Key, Aes256Key and ChaCha20Key implement Drop separately.
+impl Debug for CipherContext {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
-            SymmetricCipherKey::Aes128(_, aes_key) | SymmetricCipherKey::Aes256(_, aes_key) => unsafe {
-                #[allow(clippy::transmute_ptr_to_ptr)]
-                let value: &mut [u8; size_of::<AES_KEY>()] = transmute(aes_key);
-                value.zeroize();
+            Self::Iv128(_) => write!(f, "Iv128"),
+            Self::None => write!(f, "None"),
+        }
+    }
+}
+
+#[non_exhaustive]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+/// Cipher algorithm identifier.
```

Comment:
should we call this "block cipher", which is the standard nomenclature. But I guess all those names are just a subset of "algorithm" anyway.

Reply:
We intend for this API to potentially support Stream ciphers (like ChaCha) as well.